Notification for resolved comment

infinite-persistence opened the pull request:
Closes #7180 

_Skipped discussions -- as always, ready to iterate or even close if there is a different plan_

## Spec
- [x]  Send notification + email to:
    - [x]  **Commenter** if the comment was resolved by someone else.
    - [x]  **Repliers**
    - [x]  **Mentioned users** in both the comment and replies.
- [ ]  Notify on “un-resolve” (probably no need?)

## Decision
The biggest one is between:
1. Overloading `NotificationEventType.MentionedInDocument` to handle both "mentioned" and "resolved", vs
    - con: potentially confusing code and setting
2. Creating a new `NotificationEventType.ResolveComment`
    - con: creates another setting (too fine-grained?), migration

Went with "1" for this iteration.

tommoor commented:
I think overloading `MentionedInComment` to this extent is a bit much, it creates all sorts of logical forks and potential bugs if you get mentioned in a resolved comment thread (I don't know that this is possible today, but I wouldn't want to count it out).

Creating another event type _is_ quite fine-grained, but we can default it to enabled (no db migration required) and I think thread resolution is different enough that it makes sense. I'd rather keep the events in the backend quite specific and the UI can be changed to toggle multiple at once if necessary.

infinite-persistence commented:
Updated.

Behavioral changes in this iteration:
- The `comments.update` webhook used to include resolve, but that is now split into `comments.resolve`.

tommoor commented:
I don't really follow why it was neccessary to add the new event type? Generally trying to avoid adding new specific events as they're not backwards compatible and require all that extra plumbing that could just use `comments.update`

infinite-persistence commented:
_Thanks for patience._

Updated:
- Only split `NotificationEventType`, not `CommentUpdateEvent` (you're right).
- Fixed logic in previous "New comment ↓" hint.

tommoor commented on app/scenes/Settings/Notifications.tsx:
```diff
@@ -65,6 +66,14 @@ function Notifications() {
+      icon: <DoneIcon />,
+      title: t("Resolved"),
+      description: t(
+        "Receive a notification when someone resolves a comment you created or were involved in"
```
```suggestion
        "Receive a notification when a comment thread you were involved in is resolved"
```

tommoor commented on server/emails/templates/CommentResolvedEmail.tsx:
```diff
@@ -0,0 +1,163 @@
+  }
+
+  protected preview({ actorName }: Props): string {
+    return `${actorName} resolved a comment you were involved in`;
```
```suggestion
    return `${actorName} resolved a comment thread`;
```

tommoor commented on server/emails/templates/CommentResolvedEmail.tsx:
```diff
@@ -0,0 +1,163 @@
+  }
+
+  protected subject({ document }: Props) {
+    return `Resolved a comment in “${document.title}”`;
```
```suggestion
    return `Resolved a comment thread in “${document.title}”`;
```

tommoor commented on server/emails/templates/CommentResolvedEmail.tsx:
```diff
@@ -0,0 +1,163 @@
+    commentId,
+    collection,
+  }: Props): string {
+    const t1 = `${actorName} resolved a comment on "${document.title}"`;
```
```suggestion
    const t1 = `${actorName} resolved a comment thread on "${document.title}"`;
```

tommoor commented on server/queues/tasks/CommentUpdatedNotificationsTask.ts:
```diff
@@ -53,6 +65,86 @@ export default class CommentUpdatedNotificationsTask extends BaseTask<CommentEve
+        }
+      }
+
+      // -- mentions
```
Can we refactor this loop so that mentions is first. We can parse the mentioned user ID's, append the `createdById` to that array and then avoid having the rest of the logic twice here and the indirection to the `send` method.

infinite-persistence replied:
:flushed:

tommoor replied:
I think it's an improvement? Less bug oppos

infinite-persistence replied:
Ah, I meant to say "I'm embarrassed for not doing that sooner" through that emoji 😄 .  So much cleaner.
I was fixated at retaining the copied logic from the mentions for some reason, and only factored out `send` :(

tommoor reviewed (commented):
so close 👌
